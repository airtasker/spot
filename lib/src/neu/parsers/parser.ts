import * as path from "path";
import { CompilerOptions, Project, SourceFile, ts } from "ts-morph";
import { ApiConfig } from "../../syntax/api";
import { Contract, Endpoint, Security } from "../definitions";
import { LociTable } from "../locations";
import { TypeTable } from "../types";
import { parseEndpoint } from "./endpoint-parser";
import {
  getClassWithDecoratorOrThrow,
  getDecoratorConfigOrThrow,
  getJsDoc,
  getObjLiteralPropOrThrow,
  getPropertyWithDecorator,
  getPropValueAsStringOrThrow,
  getSelfAndLocalDependencies
} from "./parser-helpers";
import { parseSecurityHeader } from "./security-header-parser";

export function parse(sourcePath: string): Contract {
  const project = createProject();

  // Add all dependent files that the project requires
  const sourceFile = project.addExistingSourceFile(sourcePath);
  project.resolveSourceFileDependencies();

  // Validate that the project has no TypeScript syntax errors
  validateProject(project);

  return parseRootSourceFile(sourceFile);
}

/**
 * Parse a root source file to return a contract.
 */
function parseRootSourceFile(file: SourceFile): Contract {
  const lociTable = new LociTable();
  const typeTable = new TypeTable();

  const klass = getClassWithDecoratorOrThrow(file, "api");
  const decorator = klass.getDecoratorOrThrow("api");
  const decoratorConfig = getDecoratorConfigOrThrow(decorator);
  const nameProp = getObjLiteralPropOrThrow<ApiConfig>(decoratorConfig, "name");
  const nameLiteral = getPropValueAsStringOrThrow(nameProp);
  const descriptionDoc = getJsDoc(klass);

  const securityHeaderProp = getPropertyWithDecorator(klass, "securityHeader");

  const security: Security = {
    header:
      securityHeaderProp &&
      parseSecurityHeader(securityHeaderProp, typeTable, lociTable)
  };

  // Add location data
  lociTable.addMorphNode(LociTable.apiClassKey(), klass);
  lociTable.addMorphNode(LociTable.apiDecoratorKey(), decorator);
  lociTable.addMorphNode(LociTable.apiNameKey(), nameProp);
  if (descriptionDoc) {
    lociTable.addMorphNode(LociTable.apiDescriptionKey(), descriptionDoc);
  }

  // Resolve all related files
  const projectFiles = getSelfAndLocalDependencies(file);

  // Parse all endpoints
  const endpoints = projectFiles
    .reduce<Endpoint[]>(
      (acc, currentFile) =>
        acc.concat(
          currentFile
            .getClasses()
            .filter(k => k.getDecorator("endpoint") !== undefined)
            .map(k => parseEndpoint(k, typeTable, lociTable))
        ),
      []
    )
    .sort((a, b) => {
      if (a.name === b.name) {
        throw new Error(`Duplicate endpoint detected: ${a.name}`);
      }
      return b.name > a.name ? -1 : 1;
    });

  return {
    name: nameLiteral.getLiteralText(),
    description: descriptionDoc && descriptionDoc.getComment(),
    types: typeTable.toArray(),
    security,
    endpoints
  };
}

/**
 * Create a new project configured for Spot
 */
function createProject(): Project {
  const compilerOptions: CompilerOptions = {
    target: ts.ScriptTarget.ESNext,
    module: ts.ModuleKind.CommonJS,
    strict: true,
    noImplicitAny: true,
    strictNullChecks: true,
    strictFunctionTypes: true,
    strictPropertyInitialization: true,
    noImplicitThis: true,
    alwaysStrict: true,
    noImplicitReturns: true,
    noFallthroughCasesInSwitch: true,
    moduleResolution: ts.ModuleResolutionKind.NodeJs,
    experimentalDecorators: true,
    baseUrl: "./",
    paths: {
      "@airtasker/spot": [path.join(__dirname, "../lib")]
    }
  };

  // Creates a new typescript program in memory
  return new Project({ compilerOptions });
}

/**
 * Validate an AST project's correctness.
 *
 * @param project an AST project
 */
function validateProject(project: Project) {
  const diagnostics = project.getPreEmitDiagnostics();
  if (diagnostics.length > 0) {
    throw new Error(
      diagnostics
        .map(diagnostic => {
          const message = diagnostic.getMessageText();
          return typeof message === "string"
            ? message
            : message.getMessageText();
        })
        .join("\n")
    );
  }
}
