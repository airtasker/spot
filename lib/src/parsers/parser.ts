import path from "path";
import { CompilerOptions, Project, SourceFile, ts } from "ts-morph";
import { Locatable } from "../models/locatable";
import { ContractNode, EndpointNode } from "../models/nodes";
import { parseApi } from "./nodes/api-parser";
import { parseEndpoint } from "./nodes/endpoint-parser";
import { extractJsDocComment } from "./utilities/parser-utility";
import {
  parseInterfaceDeclaration,
  parseTypeNode
} from "./utilities/type-parser";
import {
  retrieveTypeReferencesFromEndpoints,
  uniqueReferences
} from "./utilities/type-reference-resolver";

export function parse(sourcePath: string): ContractNode {
  const compilerOptions: CompilerOptions = {
    target: ts.ScriptTarget.ESNext,
    module: ts.ModuleKind.CommonJS,
    strict: true,
    noImplicitAny: true,
    strictNullChecks: true,
    strictFunctionTypes: true,
    strictPropertyInitialization: true,
    noImplicitThis: true,
    alwaysStrict: true,
    noImplicitReturns: true,
    noFallthroughCasesInSwitch: true,
    moduleResolution: ts.ModuleResolutionKind.NodeJs,
    experimentalDecorators: true,
    baseUrl: "./",
    paths: {
      "@airtasker/spot": [path.join(__dirname, "../lib")]
    }
  };

  // Creates a new typescript program in memory
  const project = new Project({ compilerOptions });
  const sourceFile = project.addExistingSourceFile(sourcePath);

  // Add all dependent files that the project requires
  project.resolveSourceFileDependencies();

  // Validate that the project has no TypeScript syntax errors
  validateProject(project);

  return parseRootSourceFile(sourceFile, project);
}

/**
 * Parse a root source file to return a contract.
 */
function parseRootSourceFile(
  file: SourceFile,
  projectContext: Project
): ContractNode {
  const apiClasses = file
    .getClasses()
    .filter(klass => klass.getDecorator("api") !== undefined);

  if (apiClasses.length !== 1) {
    throw new Error(
      `expected class decorated with @api to be defined exactly once in root source file, found ${
        apiClasses.length
      } usages`
    );
  }
  const api = parseApi(apiClasses[0]);

  const endpoints = parseRecursively(file);

  const referenceTypes = uniqueReferences(
    retrieveTypeReferencesFromEndpoints(endpoints, projectContext)
  );

  // Construct the equivalent TypeNode for each reference type
  const types = referenceTypes.map(referenceType => {
    const file = projectContext.getSourceFileOrThrow(referenceType.location);

    const typeAlias = file.getTypeAlias(referenceType.name);
    if (typeAlias !== undefined) {
      const name = typeAlias.getName();
      const description = extractJsDocComment(typeAlias);
      const type = parseTypeNode(typeAlias.getTypeNodeOrThrow());
      return { description, name, type };
    }

    const interfaceDeclaration = file.getInterface(referenceType.name);
    if (interfaceDeclaration) {
      const name = interfaceDeclaration.getName();
      const description = extractJsDocComment(interfaceDeclaration);
      const type = parseInterfaceDeclaration(interfaceDeclaration);
      return { description, name, type };
    }

    throw new Error("unexpected type reference");
  });

  return { api, endpoints, types };
}

/**
 * Parses a file and its local imports/exports recursively, looking for endpoint definitions.
 *
 * @param file The current source file (starting with the root file where the API is defined).
 * @param visitedPaths The list of paths that were already visited, to avoid an infinite loop
 * with circular dependencies.
 */
function parseRecursively(
  file: SourceFile,
  visitedPaths: string[] = []
): Array<Locatable<EndpointNode>> {
  const filePath = file.getFilePath();
  if (visitedPaths.indexOf(filePath) !== -1) {
    // Don't visit the same file multiple times.
    return [];
  } else {
    visitedPaths.push(filePath);
  }

  const endpoints = file
    .getClasses()
    .filter(klass => klass.getDecorator("endpoint") !== undefined)
    .map(endpointClass => parseEndpoint(endpointClass));

  const importedFiles = file
    .getImportDeclarations()
    // We only care about local imports.
    .filter(i => i.getModuleSpecifierValue().startsWith("."))
    .map(myImport => myImport.getModuleSpecifierSourceFileOrThrow());

  const exportedFiles = file
    .getExportDeclarations()
    // We only care about local imports.
    .filter(e => {
      const moduleSpecifierValue = e.getModuleSpecifierValue();
      return moduleSpecifierValue && moduleSpecifierValue.startsWith(".");
    })
    .map(myImport => myImport.getModuleSpecifierSourceFileOrThrow());

  return [...importedFiles, ...exportedFiles].reduce<
    Array<Locatable<EndpointNode>>
  >(
    (endpointsAcc, currentFile) =>
      parseRecursively(currentFile, visitedPaths).concat(endpointsAcc),
    endpoints
  );
}

/**
 * Validate an AST project's correctness.
 *
 * @param project an AST project
 */
function validateProject(project: Project) {
  const diagnostics = project.getPreEmitDiagnostics();
  if (diagnostics.length > 0) {
    throw new Error(
      diagnostics
        .map(diagnostic => {
          const message = diagnostic.getMessageText();
          return typeof message === "string"
            ? message
            : message.getMessageText();
        })
        .join("\n")
    );
  }
}
