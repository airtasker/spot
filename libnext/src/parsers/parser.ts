import { CompilerOptions, Project, SourceFile, ts } from "ts-simple-ast";
import { ContractNode, EndpointNode } from "../models/nodes";
import { ReferenceType } from "../models/types";
import { parseApi } from "./nodes/api-parser";
import { parseEndpoint } from "./nodes/endpoint-parser";
import { extractJsDocComment } from "./utilities/parser-utility";
import { parseInterfaceDeclaration, parseType } from "./utilities/type-parser";
import {
  retrieveTypeReferencesFromEndpoints,
  retrieveTypeReferencesFromType,
  uniqueReferences
} from "./utilities/type-reference-resolver";

export function parseFilePath(
  sourcePath: string,
  customCompilerOptions: CompilerOptions = {}
): ContractNode {
  const defaultCompilerOptions: CompilerOptions = {
    target: ts.ScriptTarget.ESNext,
    module: ts.ModuleKind.CommonJS,
    strict: true,
    noImplicitAny: true,
    strictNullChecks: true,
    strictFunctionTypes: true,
    strictPropertyInitialization: true,
    noImplicitThis: true,
    alwaysStrict: true,
    noImplicitReturns: true,
    noFallthroughCasesInSwitch: true,
    moduleResolution: ts.ModuleResolutionKind.NodeJs,
    experimentalDecorators: true
  };

  const compilerOptions = {
    ...defaultCompilerOptions,
    ...customCompilerOptions
  };

  // Creates a new typescript program in memory
  const project = new Project({ compilerOptions });
  const sourceFile = project.addExistingSourceFile(sourcePath);

  // Add all dependent files that the project requires
  project.resolveSourceFileDependencies();

  // Validate that the project has no TypeScript syntax errors
  validateProject(project);

  return parseRootSourceFile(sourceFile, project);
}

/**
 * Parse a root source file to return a contract.
 */
function parseRootSourceFile(
  file: SourceFile,
  projectContext: Project
): ContractNode {
  const apiClasses = file
    .getClasses()
    .filter(klass => klass.getDecorator("api") !== undefined);

  if (apiClasses.length !== 1) {
    throw new Error(
      `expected class decorated with @api to be defined exactly once in root source file, found ${
        apiClasses.length
      } usages`
    );
  }
  const api = parseApi(apiClasses[0]);

  const sourceEndpoints = file
    .getClasses()
    .filter(klass => klass.getDecorator("endpoint") !== undefined)
    .map(endpointClass => parseEndpoint(endpointClass));

  const importedFiles = file
    .getImportDeclarations()
    .map(myImport => myImport.getModuleSpecifierSourceFileOrThrow());

  const endpoints = importedFiles.reduce<EndpointNode[]>(
    (endpointsAcc, currentFile) =>
      currentFile
        .getClasses()
        .filter(klass => klass.getDecorator("endpoint") !== undefined)
        .map(endpointClass => parseEndpoint(endpointClass))
        .concat(endpointsAcc),
    sourceEndpoints
  );

  // Direct reference types, filtered to unique references for efficiency
  const uniqueDirectReferenceTypes = uniqueReferences(
    retrieveTypeReferencesFromEndpoints(endpoints)
  );

  // Reference types from the direct reference type hierarchy
  const secondaryReferenceTypes = uniqueDirectReferenceTypes.reduce<
    ReferenceType[]
  >(
    (referenceTypesAcc, currentReferenceType) =>
      referenceTypesAcc.concat(
        retrieveTypeReferencesFromType(currentReferenceType, projectContext)
      ),
    []
  );

  // Combine and filter to unique type references
  const allReferenceTypes = uniqueReferences(
    uniqueDirectReferenceTypes.concat(secondaryReferenceTypes)
  );

  // Construct the equivalent TypeNode for each reference type
  const types = allReferenceTypes.map(referenceType => {
    const file = projectContext.getSourceFileOrThrow(referenceType.location);

    const typeAlias = file.getTypeAlias(referenceType.name);
    if (typeAlias !== undefined) {
      const name = typeAlias.getName();
      const description = extractJsDocComment(typeAlias);
      const type = parseType(typeAlias.getTypeNodeOrThrow());
      return { description, name, type };
    }

    const interfaceDeclaration = file.getInterface(referenceType.name);
    if (interfaceDeclaration) {
      const name = interfaceDeclaration.getName();
      const description = extractJsDocComment(interfaceDeclaration);
      const type = parseInterfaceDeclaration(interfaceDeclaration);
      return { description, name, type };
    }

    throw new Error("unexpected type reference");
  });

  return { api, endpoints, types };
}

/**
 * Validate an AST project's correctness.
 *
 * @param project an AST project
 */
function validateProject(project: Project) {
  const diagnostics = project.getPreEmitDiagnostics();
  if (diagnostics.length > 0) {
    throw new Error(
      diagnostics.map(diagnostic => diagnostic.getMessageText()).join("\n")
    );
  }
}
