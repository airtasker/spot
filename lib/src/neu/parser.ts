import * as path from "path";
import { CompilerOptions, Project, ts } from "ts-morph";
import { Contract } from "./definitions";
import { parseContract } from "./parsers/contract-parser";

export function parse(sourcePath: string): Contract {
  const project = createProject();

  // Add all dependent files that the project requires
  const sourceFile = project.addExistingSourceFile(sourcePath);
  project.resolveSourceFileDependencies();

  // Validate that the project has no TypeScript syntax errors
  validateProject(project);

  const result = parseContract(sourceFile);

  if (result.isErr()) throw result.unwrapErr();

  return result.unwrap().contract;
}

/**
 * Create a new project configured for Spot
 */
function createProject(): Project {
  const compilerOptions: CompilerOptions = {
    target: ts.ScriptTarget.ESNext,
    module: ts.ModuleKind.CommonJS,
    strict: true,
    noImplicitAny: true,
    strictNullChecks: true,
    strictFunctionTypes: true,
    strictPropertyInitialization: true,
    noImplicitThis: true,
    alwaysStrict: true,
    noImplicitReturns: true,
    noFallthroughCasesInSwitch: true,
    moduleResolution: ts.ModuleResolutionKind.NodeJs,
    experimentalDecorators: true,
    baseUrl: "./",
    paths: {
      "@airtasker/spot": [path.join(__dirname, "../lib")]
    }
  };

  // Creates a new typescript program in memory
  return new Project({ compilerOptions });
}

/**
 * Validate an AST project's correctness.
 *
 * @param project an AST project
 */
function validateProject(project: Project) {
  const diagnostics = project.getPreEmitDiagnostics();
  if (diagnostics.length > 0) {
    throw new Error(
      diagnostics
        .map(diagnostic => {
          const message = diagnostic.getMessageText();
          return typeof message === "string"
            ? message
            : message.getMessageText();
        })
        .join("\n")
    );
  }
}
