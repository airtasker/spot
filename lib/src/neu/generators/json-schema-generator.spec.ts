import { parse } from "../parser";
import {
  arrayType,
  booleanLiteralType,
  booleanType,
  floatLiteralType,
  floatType,
  int32Type,
  int64Type,
  intLiteralType,
  nullType,
  objectType,
  referenceType,
  stringLiteralType,
  stringType,
  unionType
} from "../types";
import {
  generateJsonSchema,
  generateJsonSchemaType
} from "./json-schema-generator";

const EXAMPLE_PATH = "./lib/src/__examples__/contract.ts";

describe("JSON Schema generator", () => {
  test("produces valid code", async () => {
    const contract = parse(EXAMPLE_PATH);
    expect(generateJsonSchema(contract, "json")).toMatchSnapshot("json");
    expect(generateJsonSchema(contract, "yaml")).toMatchSnapshot("yaml");
  });

  describe("generates type validator", () => {
    test("null", () => {
      expect(generateJsonSchemaType(nullType())).toMatchInlineSnapshot(`
        Object {
          "type": "null",
        }
      `);
    });

    test("boolean", () => {
      expect(generateJsonSchemaType(booleanType())).toMatchInlineSnapshot(`
        Object {
          "type": "boolean",
        }
      `);
    });

    test("boolean literal", () => {
      expect(generateJsonSchemaType(booleanLiteralType(true)))
        .toMatchInlineSnapshot(`
        Object {
          "const": true,
          "type": "boolean",
        }
      `);
      expect(generateJsonSchemaType(booleanLiteralType(false)))
        .toMatchInlineSnapshot(`
        Object {
          "const": false,
          "type": "boolean",
        }
      `);
    });

    test("string", () => {
      expect(generateJsonSchemaType(stringType())).toMatchInlineSnapshot(`
        Object {
          "type": "string",
        }
      `);
    });

    test("string literal", () => {
      expect(generateJsonSchemaType(stringLiteralType("some literal")))
        .toMatchInlineSnapshot(`
        Object {
          "const": "some literal",
          "type": "string",
        }
      `);
    });

    test("float", () => {
      expect(generateJsonSchemaType(floatType())).toMatchInlineSnapshot(`
        Object {
          "type": "number",
        }
      `);
    });

    test("number literal", () => {
      expect(generateJsonSchemaType(floatLiteralType(1.5)))
        .toMatchInlineSnapshot(`
        Object {
          "const": 1.5,
          "type": "number",
        }
      `);
      expect(generateJsonSchemaType(floatLiteralType(-23.1)))
        .toMatchInlineSnapshot(`
        Object {
          "const": -23.1,
          "type": "number",
        }
      `);
    });

    test("int32", () => {
      expect(generateJsonSchemaType(int32Type())).toMatchInlineSnapshot(`
        Object {
          "type": "integer",
        }
      `);
    });

    test("int64", () => {
      expect(generateJsonSchemaType(int64Type())).toMatchInlineSnapshot(`
        Object {
          "type": "integer",
        }
      `);
    });

    test("number literal", () => {
      expect(generateJsonSchemaType(intLiteralType(0))).toMatchInlineSnapshot(`
        Object {
          "const": 0,
          "type": "integer",
        }
      `);
      expect(generateJsonSchemaType(intLiteralType(123)))
        .toMatchInlineSnapshot(`
        Object {
          "const": 123,
          "type": "integer",
        }
      `);
      expect(generateJsonSchemaType(intLiteralType(-1000)))
        .toMatchInlineSnapshot(`
        Object {
          "const": -1000,
          "type": "integer",
        }
      `);
    });

    test("object", () => {
      expect(generateJsonSchemaType(objectType([]))).toMatchInlineSnapshot(`
        Object {
          "additionalProperties": true,
          "properties": Object {},
          "required": Array [],
          "type": "object",
        }
      `);
      expect(
        generateJsonSchemaType(
          objectType([
            {
              name: "singleField",
              type: floatType(),
              optional: false
            }
          ])
        )
      ).toMatchInlineSnapshot(`
        Object {
          "additionalProperties": true,
          "properties": Object {
            "singleField": Object {
              "type": "number",
            },
          },
          "required": Array [
            "singleField",
          ],
          "type": "object",
        }
      `);
      expect(
        generateJsonSchemaType(
          objectType([
            {
              name: "field1",
              type: floatType(),
              optional: false
            },
            {
              name: "field2",
              type: stringType(),
              optional: false
            },
            {
              name: "field3",
              type: booleanType(),
              optional: true
            }
          ])
        )
      ).toMatchInlineSnapshot(`
        Object {
          "additionalProperties": true,
          "properties": Object {
            "field1": Object {
              "type": "number",
            },
            "field2": Object {
              "type": "string",
            },
            "field3": Object {
              "type": "boolean",
            },
          },
          "required": Array [
            "field1",
            "field2",
          ],
          "type": "object",
        }
      `);
    });

    test("array", () => {
      expect(generateJsonSchemaType(arrayType(stringType())))
        .toMatchInlineSnapshot(`
        Object {
          "items": Object {
            "type": "string",
          },
          "type": "array",
        }
      `);
    });

    test("union", () => {
      expect(
        generateJsonSchemaType(
          unionType([stringType(), floatType(), booleanType()])
        )
      ).toMatchInlineSnapshot(`
        Object {
          "oneOf": Array [
            Object {
              "type": "string",
            },
            Object {
              "type": "number",
            },
            Object {
              "type": "boolean",
            },
          ],
        }
      `);
    });

    test("type reference", () => {
      expect(generateJsonSchemaType(referenceType("OtherType")))
        .toMatchInlineSnapshot(`
        Object {
          "$ref": "#/definitions/OtherType",
        }
      `);
    });
  });
});
