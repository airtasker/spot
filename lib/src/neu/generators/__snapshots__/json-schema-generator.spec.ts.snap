// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSON Schema generator produces valid code: json 1`] = `
"{
  \\"$schema\\": \\"http://json-schema.org/draft-07/schema#\\",
  \\"definitions\\": {
    \\"Address\\": {
      \\"type\\": \\"string\\"
    },
    \\"Company\\": {
      \\"type\\": \\"object\\",
      \\"properties\\": {
        \\"id\\": {
          \\"type\\": \\"string\\"
        }
      },
      \\"required\\": [
        \\"id\\"
      ],
      \\"additionalProperties\\": true
    },
    \\"CompanyBody\\": {
      \\"type\\": \\"object\\",
      \\"properties\\": {
        \\"data\\": {
          \\"$ref\\": \\"#/definitions/Company\\"
        }
      },
      \\"required\\": [
        \\"data\\"
      ],
      \\"additionalProperties\\": true
    },
    \\"CreateUserRequestBody\\": {
      \\"type\\": \\"object\\",
      \\"properties\\": {
        \\"data\\": {
          \\"type\\": \\"object\\",
          \\"properties\\": {
            \\"firstName\\": {
              \\"type\\": \\"string\\"
            },
            \\"lastName\\": {
              \\"type\\": \\"string\\"
            },
            \\"age\\": {
              \\"type\\": \\"number\\"
            },
            \\"email\\": {
              \\"$ref\\": \\"#/definitions/Email\\"
            },
            \\"address\\": {
              \\"$ref\\": \\"#/definitions/Address\\"
            },
            \\"createdAt\\": {
              \\"type\\": \\"string\\",
              \\"format\\": \\"date\\"
            }
          },
          \\"required\\": [
            \\"firstName\\",
            \\"lastName\\",
            \\"age\\",
            \\"email\\",
            \\"address\\"
          ],
          \\"additionalProperties\\": true
        }
      },
      \\"required\\": [
        \\"data\\"
      ],
      \\"additionalProperties\\": true
    },
    \\"Email\\": {
      \\"type\\": \\"string\\"
    },
    \\"ErrorBody\\": {
      \\"type\\": \\"object\\",
      \\"properties\\": {
        \\"name\\": {
          \\"type\\": \\"string\\"
        },
        \\"message\\": {
          \\"type\\": \\"array\\",
          \\"items\\": {
            \\"type\\": \\"string\\"
          }
        }
      },
      \\"required\\": [
        \\"name\\",
        \\"message\\"
      ],
      \\"additionalProperties\\": true
    },
    \\"MessageOptions\\": {
      \\"type\\": \\"object\\",
      \\"properties\\": {
        \\"newsletter\\": {
          \\"type\\": \\"boolean\\"
        }
      },
      \\"required\\": [
        \\"newsletter\\"
      ],
      \\"additionalProperties\\": true
    },
    \\"Profile\\": {
      \\"type\\": \\"object\\",
      \\"properties\\": {
        \\"private\\": {
          \\"type\\": \\"boolean\\"
        },
        \\"messageOptions\\": {
          \\"$ref\\": \\"#/definitions/MessageOptions\\"
        }
      },
      \\"required\\": [
        \\"private\\",
        \\"messageOptions\\"
      ],
      \\"additionalProperties\\": true
    },
    \\"UserBody\\": {
      \\"type\\": \\"object\\",
      \\"properties\\": {
        \\"data\\": {
          \\"type\\": \\"object\\",
          \\"properties\\": {
            \\"firstName\\": {
              \\"type\\": \\"string\\"
            },
            \\"lastName\\": {
              \\"type\\": \\"string\\"
            },
            \\"profile\\": {
              \\"$ref\\": \\"#/definitions/Profile\\"
            }
          },
          \\"required\\": [
            \\"firstName\\",
            \\"lastName\\",
            \\"profile\\"
          ],
          \\"additionalProperties\\": true
        }
      },
      \\"required\\": [
        \\"data\\"
      ],
      \\"additionalProperties\\": true
    },
    \\"UserQuery\\": {
      \\"type\\": \\"object\\",
      \\"properties\\": {
        \\"id\\": {
          \\"type\\": \\"number\\"
        },
        \\"slug\\": {
          \\"type\\": \\"string\\"
        }
      },
      \\"required\\": [
        \\"id\\",
        \\"slug\\"
      ],
      \\"additionalProperties\\": true
    }
  }
}"
`;

exports[`JSON Schema generator produces valid code: yaml 1`] = `
"$schema: 'http://json-schema.org/draft-07/schema#'
definitions:
  Address:
    type: string
  Company:
    type: object
    properties:
      id:
        type: string
    required:
      - id
    additionalProperties: true
  CompanyBody:
    type: object
    properties:
      data:
        $ref: '#/definitions/Company'
    required:
      - data
    additionalProperties: true
  CreateUserRequestBody:
    type: object
    properties:
      data:
        type: object
        properties:
          firstName:
            type: string
          lastName:
            type: string
          age:
            type: number
          email:
            $ref: '#/definitions/Email'
          address:
            $ref: '#/definitions/Address'
          createdAt:
            type: string
            format: date
        required:
          - firstName
          - lastName
          - age
          - email
          - address
        additionalProperties: true
    required:
      - data
    additionalProperties: true
  Email:
    type: string
  ErrorBody:
    type: object
    properties:
      name:
        type: string
      message:
        type: array
        items:
          type: string
    required:
      - name
      - message
    additionalProperties: true
  MessageOptions:
    type: object
    properties:
      newsletter:
        type: boolean
    required:
      - newsletter
    additionalProperties: true
  Profile:
    type: object
    properties:
      private:
        type: boolean
      messageOptions:
        $ref: '#/definitions/MessageOptions'
    required:
      - private
      - messageOptions
    additionalProperties: true
  UserBody:
    type: object
    properties:
      data:
        type: object
        properties:
          firstName:
            type: string
          lastName:
            type: string
          profile:
            $ref: '#/definitions/Profile'
        required:
          - firstName
          - lastName
          - profile
        additionalProperties: true
    required:
      - data
    additionalProperties: true
  UserQuery:
    type: object
    properties:
      id:
        type: number
      slug:
        type: string
    required:
      - id
      - slug
    additionalProperties: true
"
`;
